src/dataloader.py

```diff
@@ -0,0 +1,90 @@
+import itertools
+import os
+from dataclasses import dataclass, field
+from typing import Dict, Literal, Optional
+
+import torch
+from datasets import IterableDatasetDict, concatenate_datasets, load_dataset
+from torch.utils.data import DataLoader
+from transformers import DataCollatorForLanguageModeling, PreTrainedTokenizerBase
+
+
+@dataclass
+class BooksCorpusAndWiki:
+    tokenizer: PreTrainedTokenizerBase
+    # stream shuffled dataset by batches of buffer_size
+    buffer_size: int = 10000
+    max_seq_length: int = 128
+    batch_size: Dict[str, int] = field(default_factory=lambda: {"train": 32, "validation": 32})
+    # % tokens to mask for Encoders. Set to 0% to disable for Decoders.
+    mlm_probability: float = 0.15
+    # random seed for shuffling and masking.
+    seed: int = 42
+    # HuggingFace iterable dataset dict.
+    datasets: Optional[IterableDatasetDict] = None
+    # no. cpu threads. Defaults to # cores - 2.
+    num_workers: int = os.cpu_count() - 2
+
+    def setup(self, suffix=""):
+        """Streams, shuffles, tokenizes, and groups the datasets.
+
+        Args:
+            suffix (str, optional): % of datasplit. Defaults to 100%. Format: [:n%]
+        """
+
+        # stream the dataset as it is too large to download
+        bookcorpus = load_dataset("bookcorpus", split=f"train{suffix}", streaming=True)
+        wiki_train = load_dataset("wiki40b", "en", split=f"train{suffix}", streaming=True)
+        wiki_val = load_dataset("wiki40b", "en", split=f"validation{suffix}", streaming=True)
+
+        # keep text column to concat with bookcorpus
+        wiki_train = wiki_train.remove_columns([col for col in wiki_train.column_names if col != "text"])
+        wiki_val = wiki_val.remove_columns([col for col in wiki_val.column_names if col != "text"])
+
+        assert bookcorpus.features.type == wiki_train.features.type
+
+        # dataset transform steps
+        self.datasets = (
+            IterableDatasetDict({"train": concatenate_datasets([bookcorpus, wiki_train]), "validation": wiki_val})
```

Comment:
Should check streaming perf -- when connected to model. Is `streaming`'s dynamic JIT data processing a bottleneck?

Reply:
from my limited experience. yeah. streaming datasets is a huge bottleneck unless we parallelize it to stream future batches on another thread while concurrently processing the current batch. I did a quick google search and it doesn't seem too hard to implement that https://discuss.huggingface.co/t/streaming-dataset-and-cache/32751/5